README.md
# Rack DigitalOcean AppPlatform Safelist

Simple rack middleware for ruby applications hosted on Digital Ocean app platform to block ip addresses that are not
safelisted via an environment variable.

Simply add an environment variable called ALLOWED_IPS that contains a comma seperated list of ips that are allowed to
access you application (you can use CIDR notation as well to safelist ips).

Example:

ALLOWED_IPS = "8.8.8.8, 8.8.4.4, 1.2.3.0/24"

## Getting started

### Installing

Add this line to your application's Gemfile:

```ruby
# In your Gemfile

gem "rack_do_app_platform_safelist"
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rack_do_app_platform_safelist

### Plugging into the application

Then tell your ruby web application to use the middleware.

a) For __rails__ applications.

```ruby
# In config/initializers/rack_do_app_platform_safelist.rb

require "rack_do_app_platform_safelist"

if Rails.env.production?
  Rails.application.config.middleware.insert_before(0, RackDoAppPlatformSafelist, logger: Rails.logger)
end
```

This will insert the middleware at the top of the middleware stack so it can block request before reaching any other
middleware in your application. You would only want to enable this middleware in production.

b) For __rack__ applications:

```ruby
# In config.ru

require "rack_do_app_platform_safelist"

use RackDoAppPlatformSafelist
run App.new
```

### Options

If you want to use a different environment variable for the safelisted IPs you can pass it as an argument

a) For __rails__ applications.

```ruby
# In config/initializers/rack_do_app_platform_safelist.rb

require "rack_do_app_platform_safelist"

Rails.application.config.middleware.insert_before(0, RackDoAppPlatformSafelist, logger: Rails.logger,
                                                                                env_key: "SAFELISTED_IPS")
```

b) For __rack__ applications:

```ruby
# In config.ru

require "rack_do_app_platform_safelist"

use RackDoAppPlatformSafelist, env_key: "SAFELISTED_IPS"
run App.new
```

The middleware will now look for the IP addresses to safelist in the SAFELISTED_IPS environment variable.


CHANGELOG.md
## [Unreleased]

- Updated rubocop 1.50.2 -> 1.52.0
- Updated debug 1.7.2 -> 1.8.0
- Updated rubocop-minitest 0.30.0 -> 0.31.0
- Updated bundler 2.4.6 -> 2.4.13

## [1.0.1] - 2023-02-12

- Changed readme to be more thorough with regards to integrating with your application.
- Fixed repository errors in gemspec.

## [1.0.0] - 2023-02-11

- Initial release


test/test_helper.rb
# frozen_string_literal: true

require "simplecov"
SimpleCov.start

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "rack_do_app_platform_safelist"
require "minitest/autorun"
require "rack"
require "rack/test"
require "debug"


test/rack_do_app_platform_safelist_test.rb
# frozen_string_literal: true

require "test_helper"

ENV["ALLOWED_IPS"] = "8.8.8.8, 1.1.1.0/24"
ENV["SAFELISTED_IPS"] = "8.8.4.4"

class RackDoAppPlatformSafelistTest < Minitest::Test
  include Rack::Test::Methods

  class TestApp
    def call(_env)
      [200, { "Content-Type" => "text/plain" }, ["Success"]]
    end
  end

  def app
    send(@rack_app || :default_rack_app)
  end

  def test_that_it_has_a_version_number
    refute_nil ::RackDoAppPlatformSafelist::VERSION
  end

  def test_safelisted_ips_are_not_blocked
    header "X-Forwarded-for", "8.8.8.8,10.0.0.1"
    get "/"

    assert_same(200, last_response.status)
  end

  def test_safelisted_ip_ranges_are_not_blocked
    header "X-Forwarded-for", "1.1.1.11,10.0.0.1"
    get "/"

    assert_same(200, last_response.status)
  end

  def test_non_safelisted_ips_are_blocked
    header "X-Forwarded-for", "8.8.4.4,10.0.0.1"
    get "/"

    assert_same(403, last_response.status)
  end

  def test_non_safelisted_ips_are_logged
    @logger = Minitest::Mock.new
    @logger.expect(:info, true, ["Blocking request - IP address 8.8.4.4 is not safelisted"])
    header "X-Forwarded-for", "8.8.4.4,10.0.0.1"
    get "/"

    assert(@logger.verify)
  end

  def test_log_is_created_if_client_ip_can_not_be_determiend
    @logger = Minitest::Mock.new
    @logger.expect(:info, true, ["Blocking request - Could not determine client ip address"])
    header "X-Forwarded-for", "10.0.0.1"
    get "/"

    assert(@logger.verify)
  end

  def test_log_is_created_if_forwarded_for_header_is_not_prensent
    @logger = Minitest::Mock.new
    @logger.expect(:info, true, ["Blocking request - Could not determine client ip address"])
    get "/"

    assert(@logger.verify)
  end

  def test_you_can_pass_in_env_variable_to_use
    @rack_app = :rack_app_with_custom_env_variable
    header "X-Forwarded-for", "8.8.4.4,10.0.0.1"
    get "/"

    assert_same(200, last_response.status)
  end

  def test_rack_logger_is_used_if_present_in_env
    logger = Minitest::Mock.new
    logger.expect(:info, true, ["Blocking request - IP address 8.8.4.4 is not safelisted"])
    header "X-Forwarded-for", "8.8.4.4,10.0.0.1"
    get "/", nil, "rack.logger" => logger

    assert(logger.verify)
  end

  private

  def default_rack_app
    logger = @logger

    Rack::Builder.new do
      use RackDoAppPlatformSafelist, logger: logger
      run TestApp.new
    end
  end

  def rack_app_with_custom_env_variable
    logger = @logger

    Rack::Builder.new do
      use RackDoAppPlatformSafelist, logger: logger, env_key: "SAFELISTED_IPS"
      run TestApp.new
    end
  end
end


test/test_rack_do_app_platform_safelist.rb
# frozen_string_literal: true

require "test_helper"

class TestRackDoAppPlatformSafelist < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::RackDoAppPlatformSafelist::VERSION
  end

  def test_it_does_something_useful
    assert false
  end
end


lib/rack_do_app_platform_safelist.rb
# frozen_string_literal: true

require "ipaddr"
require "logger"

require_relative "rack_do_app_platform_safelist/version"

# Rack middleware for digital ocean app platform that will block any ip address that is not saflisted using the
# ALLOWED_IPS enviroment variable by default.
#
#  Eg:
#   ALLOWED_IPS = "8.8.8.8,8.8.4.4,1.2.3.0/24"
class RackDoAppPlatformSafelist
  attr_reader :allowed_ips

  # @param logger [#info] the logger to use for messages. If nil it will use the value set in "rack.logger" if present
  #   or create a new logger instance.
  # @param env_key [#to_s] the environment variable to use to get the safelisted ips.
  def initialize(app, logger: nil, env_key: "ALLOWED_IPS")
    @app = app
    @logger = logger
    @allowed_ips = ENV.fetch(env_key.to_s, "").split(",").map(&:strip).uniq.map { |ip| IPAddr.new(ip) }
  end

  def call(env)
    client_ip = extract_client_ip(env)
    return @app.call(env) if safelisted_ip?(client_ip)

    logger(env).info(error_message(client_ip))
    [403, { "content-type" => "text/plain" }, ["Forbidden\n"]]
  end

  private

  def safelisted_ip?(client_ip)
    return false if client_ip.nil?

    IPAddr.new(client_ip).then { |ip| allowed_ips.any? { |allowed_ip| allowed_ip.include?(ip) } }
  end

  def extract_client_ip(env)
    # DigitalOcean's load balancer appends the client ip and the load balancer's ip to the X-Forwarded-For header so
    # since we know there will always be one load balancer infront of your application on App Platform we can split the
    # string on commas and the client ip will always be the second to last item in the array.
    env["HTTP_X_FORWARDED_FOR"].to_s.split(",")[-2]
  end

  def error_message(client_ip)
    message = client_ip.nil? ? "Could not determine client ip address" : "IP address #{client_ip} is not safelisted"

    "Blocking request - #{message}"
  end

  def logger(env)
    if @logger
      @logger
    elsif env["rack.logger"]
      env["rack.logger"]
    else
      ::Logger.new(env["rack.errors"])
    end

    # return @logger if @logger
    # return env["rack.logger"] if env["rack.logger"]

    # ::Logger.new(env["rack.errors"])
  end
end
